cpp/dolfinx/la/SparsityPattern.cpp

```diff
@@ -348,35 +303,74 @@ void SparsityPattern::assemble()
       {
         // Convert to local column index
         const std::int32_t J = col - local_range1[0];
-        _diagonal_cache[row_local].push_back(J);
+        _cache_owned[row_local].push_back(J);
       }
       else
       {
-        assert(row_local < (std::int32_t)_off_diagonal_cache.size());
-        _off_diagonal_cache[row_local].push_back(col);
+        // Column index may not exist in column indexmap
+        auto it = global_to_local.insert({col, local_i});
+        if (it.second)
+        {
+          ++local_i;
+          ghosts1.push_back(col);
+          ghost_owners1.push_back(ghost_data_received[i + 2]);
+        }
+        const std::int32_t col_local = it.first->second;
+        _cache_owned[row_local].push_back(col_local);
       }
     }
   }
 
-  _diagonal_cache.resize(local_size0);
-  for (std::vector<std::int32_t>& row : _diagonal_cache)
+  // Sort and remove duplicates
+  std::vector<std::int32_t> adj_counts(local_size0, 0);
+  std::vector<std::int32_t> adj_data;
+  std::vector<std::int32_t> adj_counts_off(local_size0, 0);
+  std::vector<std::int32_t> adj_data_off;
+  for (std::int32_t i = 0; i < local_size0; ++i)
   {
+    std::vector<std::int32_t>& row = _cache_owned[i];
     std::sort(row.begin(), row.end());
-    row.erase(std::unique(row.begin(), row.end()), row.end());
+    const std::vector<std::int32_t>::iterator it_end
+        = std::unique(row.begin(), row.end());
+
+    // Find position of first "off-diagonal" column
+    const std::vector<std::int32_t>::iterator it_diag
+        = std::lower_bound(row.begin(), it_end, local_size1);
+
+    adj_data.insert(adj_data.end(), row.begin(), it_diag);
+    adj_counts[i] += (it_diag - row.begin());
+    adj_data_off.insert(adj_data_off.end(), it_diag, it_end);
+    adj_counts_off[i] += (it_end - it_diag);
+    std::vector<std::int32_t>().swap(row);
   }
-  _diagonal
-      = std::make_shared<graph::AdjacencyList<std::int32_t>>(_diagonal_cache);
-  std::vector<std::vector<std::int32_t>>().swap(_diagonal_cache);
+  std::vector<std::vector<std::int32_t>>().swap(_cache_owned);
 
-  _off_diagonal_cache.resize(local_size0);
-  for (std::vector<std::int64_t>& row : _off_diagonal_cache)
-  {
-    std::sort(row.begin(), row.end());
-    row.erase(std::unique(row.begin(), row.end()), row.end());
-  }
-  _off_diagonal = std::make_shared<graph::AdjacencyList<std::int64_t>>(
-      _off_diagonal_cache);
-  std::vector<std::vector<std::int64_t>>().swap(_off_diagonal_cache);
+  std::vector<std::int32_t> adj_offsets(local_size0 + 1);
+  std::partial_sum(adj_counts.begin(), adj_counts.end(),
+                   adj_offsets.begin() + 1);
+  std::vector<std::int32_t> adj_offsets_off(local_size0 + 1);
+  std::partial_sum(adj_counts_off.begin(), adj_counts_off.end(),
+                   adj_offsets_off.begin() + 1);
+
+  // FIXME: after transferring rows, the "correct" row map has no ghosts,
```

Comment:
This is a bit cryptic. What's the "original" l2g? I would expect this is just one.

Reply:
I'll expand the explanation here